### PR TITLE
SOLR-12490 Added parsing of json.queries for referring in JSON facets

### PR DESCRIPTION
**Pull request by a-siuniaev**
# Description

Added DSL support for queries - to have the ability to refer from domain.filter. It supports both single and multi-value query parameters. Support forexclusion is not yet implemented.

# Solution

Intoduced json.queries field, which is translated with query DSL (see [comment](https://issues.apache.org/jira/browse/SOLR-12490?focusedCommentId=16514420&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16514420)).

# Checklist

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java [hunk 1]
```diff
@@ -191,7 +191,9 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
     }
 
     // implement compat for existing components...
-    JsonQueryConverter jsonQueryConverter = new JsonQueryConverter();
+    JsonQueryConverter jsonQueryConverter = (JsonQueryConverter) req.getContext()
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: I think it's necessary only for the earlier "dynamic" approach. I suppose it's not necessary in the eager one.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java [hunk 2]
```diff
@@ -214,37 +216,22 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
           out = "rows";
         } else if (SORT.equals(key)) {
           out = SORT;
+        } else if ("queries".equals(key)) {
+          for (Map.Entry<String, Object> subEntry : ((Map<String, Object>) entry.getValue()).entrySet()) {
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: Pls check before cast, throw exception with meaningful message and assert 400 http code in test.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java [hunk 3]
```diff
@@ -214,37 +216,22 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
...
-            }
-          } else {
-            newval[newval.length-1] = isQuery ? jsonQueryConverter.toLocalParams(val, newMap) : val.toString();
-          }
-          newMap.put(out, newval);
-        } else {
-          newMap.put(out, new String[]{isQuery ? jsonQueryConverter.toLocalParams(val, newMap) : val.toString()});
-        }
-
+        processJsonEntry(newMap, jsonQueryConverter, entry, out, isQuery, arr);
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: `process` is a meaningless verb, but I saw enclosing method. ok. From JSON's POV it's a `property` not entry.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java (hunk 1 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: True. Reversed.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java (hunk 3 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: Tried my best to refactor the variables' names.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java (hunk 2 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: Done.

File: solr/core/src/java/org/apache/solr/request/json/JsonQueryConverter.java [hunk 4]
```diff
@@ -28,6 +28,10 @@
  * @lucene.internal
  */
 class JsonQueryConverter {
+  public static final String paramsPrefix = "_tt";
+
+  public static final Object contextKey = JsonQueryConverter.class.getSimpleName();
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/request/json/JsonQueryConverter.java]: Now, It's redundant, I suppose.

File: solr/core/src/java/org/apache/solr/request/json/RequestUtil.java [hunk 5]
```diff
@@ -214,48 +215,60 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
...
+          if (queriesJsonObj instanceof Map) {
+            for (Map.Entry<String, Object> queryJsonProperty : ((Map<String, Object>) queriesJsonObj).entrySet()) {
+              out = queryJsonProperty.getKey();
+              arr = true;
+              isQuery = true;
+              convertJsonPropertyToLocalParams(newMap, jsonQueryConverter, queryJsonProperty, out, isQuery, arr);
+            }
+            continue;
+          } else {
+           throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Expected Map for 'queries', received " + queriesJsonObj.getClass().getSimpleName() + "=" + queriesJsonObj);
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/request/json/RequestUtil.java]: not Map. but Object for json.

**Pull request closed without merging**